### PR TITLE
[SourceKit] Include AccessLevel attributes in structure (SR-5978)

### PR DESCRIPTION
**Pull request by marcelofabri**
This PR includes AccessLevel attributes inside `key.attributes` in structures.

This would make possible for tools like [SwiftLint](https://github.com/realm/SwiftLint) to get the ranges of these attributes, which can be useful for several rules:

* Enforcing that every top level declaration has an explicit AccessLevel
* Enforcing that all declarations have explicit AccessLevels
* Enforcing that an AccessLevel is omitted when it can be inferred
* Validating the order of "modifiers" (`private`, `mutating`, `final`, etc) of declarations
* Enforcing that the setter AccessLevel is not the same as the general one (e.g. `private(set) private foo: Int`)

This PR also adds attributes for setter AccessLevel in 472fd78c43203a6c746db720373c79352a066a29. I'm not sure if this is the best way to do it, creating attributes like `source.decl.attribute.setter_access.private`, so feedbacks are welcome!

Resolves [SR-5978](https://bugs.swift.org/browse/SR-5978).

// cc @nkcsgexi

### CODE REVIEW

File: tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp [hunk 1]
```diff
@@ -623,10 +623,47 @@ Optional<UIdent> SwiftLangSupport::getUIDForDeclAttribute(const swift::DeclAttri
         return Attr_Objc;
       }
     }
+    case DAK_AccessControl: {
+      static UIdent Attr_Private("source.decl.attribute.private");
+      static UIdent Attr_FilePrivate("source.decl.attribute.fileprivate");
+      static UIdent Attr_Internal("source.decl.attribute.internal");
+      static UIdent Attr_Public("source.decl.attribute.public");
+      static UIdent Attr_Open("source.decl.attribute.open");
```
Comment (nkcsgexi) [inline tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp]: Any reason we are not adding these to ProtocolUIDs.def?

File: tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp [hunk 2]
```diff
@@ -623,10 +623,47 @@ Optional<UIdent> SwiftLangSupport::getUIDForDeclAttribute(const swift::DeclAttri
...
+        case AccessLevel::Open:
+          return Attr_Open;
+      }
+    }
+    case DAK_SetterAccess: {
+      static UIdent Attr_Private("source.decl.attribute.setter_access.private");
+      static UIdent Attr_FilePrivate("source.decl.attribute.setter_access.fileprivate");
+      static UIdent Attr_Internal("source.decl.attribute.setter_access.internal");
+      static UIdent Attr_Public("source.decl.attribute.setter_access.public");
+      static UIdent Attr_Open("source.decl.attribute.setter_access.open");
```
Comment (nkcsgexi) [inline tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp]: Same for these new keys.

File: tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp [hunk 3]
```diff
@@ -623,10 +623,47 @@ Optional<UIdent> SwiftLangSupport::getUIDForDeclAttribute(const swift::DeclAttri
...
+
+      switch (cast<AbstractAccessControlAttr>(Attr)->getAccess()) {
```
Comment (nkcsgexi) [inline tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp]: Is there a way we can share the two switch statements? It seems to me a logic duplication.

**Review by nkcsgexi — CHANGES_REQUESTED** [review]

File: tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp (hunk 1 above)
Comment (marcelofabri) [reply tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp]: Attributes are defined in `Attr.def`. The way that AccessLevel attributes work is via `DECL_ATTR_ALIAS`, which forces us to explicitly check the AccessLevel and pick the right UID.

File: tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp (hunk 3 above)
Comment (marcelofabri) [reply tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp]: I agree that there's some logic duplication, but I can't think of a way of sharing those and being easier to read at the same time.

File: tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp (hunk 1 above)
Comment (nkcsgexi) [reply tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp]: ah, OK. Thanks for the explanation!

**Review by nkcsgexi — APPROVED** [review]

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by marcelofabri** [comment]
@nkcsgexi can you please trigger @swift-ci again?

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by marcelofabri** [comment]
ping @nkcsgexi

**Pull request merged**